retrievalmarket/impl/provider_environments.go

```diff
@@ -134,46 +125,17 @@ func (pde *providerDealEnvironment) Node() retrievalmarket.RetrievalProviderNode
 	return pde.p.node
 }
 
-func (pde *providerDealEnvironment) ReadIntoBlockstore(storeID multistore.StoreID, pieceData io.ReadCloser) error {
-	// Get the the destination multistore
-	store, loadErr := pde.p.multiStore.Get(storeID)
-	if loadErr != nil {
-		return xerrors.Errorf("failed to read file into blockstore: failed to get multistore %d: %w", storeID, loadErr)
-	}
-
-	// Load the CAR into the blockstore
-	_, loadErr = cario.NewCarIO().LoadCar(store.Bstore, pieceData)
-	if loadErr != nil {
-		// Just log the error, so we can drain and close the reader before
-		// returning the error
-		loadErr = xerrors.Errorf("failed to load car file into blockstore: %w", loadErr)
-		log.Error(loadErr.Error())
-	}
-
-	// Attempt to drain and close the reader before returning any error
-	_, drainErr := io.Copy(ioutil.Discard, pieceData)
-	closeErr := pieceData.Close()
-
-	// If there was an error loading the CAR file into the blockstore, throw that error
-	if loadErr != nil {
-		return loadErr
-	}
-
-	// If there was an error draining the reader, throw that error
-	if drainErr != nil {
-		err := xerrors.Errorf("failed to read file into blockstore: failed to drain piece reader: %w", drainErr)
-		log.Error(err.Error())
-		return err
-	}
-
-	// If there was an error closing the reader, throw that error
-	if closeErr != nil {
-		err := xerrors.Errorf("failed to read file into blockstore: failed to close reader: %w", closeErr)
-		log.Error(err.Error())
-		return err
+// PrepareBlockstore is called when the deal data has been unsealed and we need
+// to add all blocks to a blockstore that is used to serve retrieval
+func (pde *providerDealEnvironment) PrepareBlockstore(ctx context.Context, dealID retrievalmarket.DealID, pieceCid cid.Cid) error {
+	// Load the blockstore that has the deal data
+	bs, err := pde.p.dagStore.LoadShard(ctx, pieceCid)
```

Comment:
Checking: does this execute in its own goroutine, because it might block for a long time while the shard is being fetched and indexed?

Reply:
Yes, this executes in the "per-deal" go-routine that the Markets fsm spins up and will NOT block other deals.